src/pybind/mgr/dashboard/frontend/src/app/shared/components/cd-label/cd-label.component.ts

```diff
@@ -0,0 +1,76 @@
+import { Component, Input, OnInit } from '@angular/core';
+
+@Component({
+  selector: 'cd-label',
+  templateUrl: './cd-label.component.html',
+  styleUrls: ['./cd-label.component.scss']
+})
+export class CdLabelComponent implements OnInit {
+  @Input() text: string;
+  // element key, for key-value elements
+  @Input() key?: string;
+  // element value, for key-value elements
+  @Input() value?: string;
+  // map elements to predefined css classes
+  @Input() classMap?: string[];
+  // property to hash the keys, default is set to hash text
+  @Input() hashKey?: any;
+  isHashingKey: boolean;
+  // property to hash the values, default is set to hash text
+  @Input() hashValue?: any;
+  isHashingValue: boolean;
+
+  className: string;
+
+  readonly cssClasses: string[] = [
+    'cd-label-green',
+    'cd-label-cyan',
+    'cd-label-purple',
+    'cd-label-light-blue',
+    'cd-label-gold',
+    'cd-label-light-green'
+  ];
+
+  ngOnInit() {
+    this.isHashingKey = this.hashKey !== undefined;
+    this.isHashingValue = this.hashValue !== undefined;
+    this.classMapToColorClass();
```

Comment:
I see an issue with the current approach. The color will be set once (on ngInit) and will remain like that even if there are changes to the value argument.

With Angular it's extremely easy to make components react to input changes, but you need to design the component starting from the template, rather than from the ts file (classic OOP approach).

Right now I see too many different use cases here, which probably complicate the design and implementation.

Let's just focus on the simplest use case, the user providing the value-class map. Ok, right now I'm realizing that for that use case we don't even need a specific component 🤦🏼 Something like the following would do the trick:
```html
<span *ngFor="let service of host.services" class="badge badge-service-{{service}}" >service: {{ service }}</span>
```
After that we only need to go to the global style sheet and properly define the `badge-services-*` for the services that we want to look different:
```css
.badge-service-osd {
  background-color: red;
}
```

We could try to encapsulate some of the logic above, but I'm not sure if we really gain much:
```html
<cd-label *ngFor="let service of host.services" key="service" [value]="service"></cd-label>
```
And that would require a template like this (it's a logic-less component, no ts file):
```html
<span class="badge badge-{{ key }}-{{ value }}" >{{ key }}: {{ value }}</span>
```
Is it worth the effort? What do you think?

---

For the rest of the use cases (hash, etc.) I'd take a similar approach (for example, given that you want to generate a color from the value argument, something like a pipe would be the optimal approach here):
```html
<span class="badge" [ngStyle]="{background-color: value | colorFromText}" >{{ key }}: {{ value }}</span>
```

Reply:
>I see an issue with the current approach. The color will be set once (on ngInit) and will remain like that even if there are changes to the value argument.

But If the component where is being used updates the data on it's own, it would already rerender and update everything that contains, right? Meaning that it would trigger this component's ngOnInit() function and update the color. 

>Right now I see too many different use cases here, which probably complicate the design and implementation.

Completly agree, there are too many different use cases, but that doesn't mean that you need to know everything the component can do, it just means that all the posible needs are covered within the same component. You could keep it as simple as just setting up the text and everything will be taken care of by the component.

>Ok, right now I'm realizing that for that use case we don't even need a specific component 🤦🏼 Something like the following would do the trick

You are right, we wouldn't even need a new component to achive the simple styling for the basic use cases we were trying to achive.

>After that we only need to go to the global style sheet and properly define the badge-services-* for the services that we want to look different:

However with this, the developer would still need to think about the colors, the color scheme and the appropriate colors within bootstrap theme and patternfly that they need, which could cause the use of inappropiate colors like red, green, yellow, etc... That already have specific use cases. And also they would have to modify the style sheets everytime which could lead to wrongly editing locally style sheets which in the end would make a mess. So I am not very much in favor of this approach

> And that would require a template like this (it's a logic-less component, no ts file):

>`<span class="badge badge-{{ key }}-{{ value }}" >{{ key }}: {{ value }}</span>`

I am not really sure what are we achiving with this. Just a very basic use case for key-value pairs, we would need to have different scss classes for all the possible values? I don't see how this would really work. 

After many different comments and approaches to different problems along the way I think right now we should reevaluate what we want to accomplish with this PR/component since, in my opinion we are not going anywhere with it since within every change we are trying to go with a new approach to the initial problem/solution.

Please @epuertat let me know what you think. Thank you.

Reply:
> > I see an issue with the current approach. The color will be set once (on ngInit) and will remain like that even if there are changes to the value argument.
> 
> But If the component where is being used updates the data on it's own, it would already rerender and update everything that contains, right? Meaning that it would trigger this component's ngOnInit() function and update the color.

No, the ngOnInit is [only triggered once](https://angular.io/guide/lifecycle-hooks) after the class constructor. If you want to respond to changes, you have to implement the ngOnChanges method. In any case, there's usually a better way to respond to changes (in fact Angular is designed to do that by default, but you have to use it correctly). Usually the best way is to expose (bind) the changing properties in the HTML template. IMHO manually tweaking the ngOnChanges is often (not always) a smell of overengineering.

> > Right now I see too many different use cases here, which probably complicate the design and implementation.
> 
> Completly agree, there are too many different use cases, but that doesn't mean that you need to know everything the component can do, it just means that all the posible needs are covered within the same component. You could keep it as simple as just setting up the text and everything will be taken care of by the component.

Yes, that's ok as long as the implementation and side effects are kept simple... However, when a helper becomes a Swiss-army knife that's usually not the case anymore.

> > Ok, right now I'm realizing that for that use case we don't even need a specific component 🤦🏼 Something like the following would do the trick
> 
> You are right, we wouldn't even need a new component to achive the simple styling for the basic use cases we were trying to achive.
> 
> > After that we only need to go to the global style sheet and properly define the badge-services-* for the services that we want to look different:
> 
> However with this, the developer would still need to think about the colors, the color scheme and the appropriate colors within bootstrap theme and patternfly that they need, which could cause the use of inappropiate colors like red, green, yellow, etc... That already have specific use cases. And also they would have to modify the style sheets everytime which could lead to wrongly editing locally style sheets which in the end would make a mess. So I am not very much in favor of this approach

After several iterations discussing this my view is that what we need is something that, by default, gives the standard `badge` look and feel, but allows to give specific colors (beyond the `badge-*` collection) to certain labels (e.g.: `service: osd`, etc). In those cases, probably we do want to choose the colors (in fact, we were already doing that with the `-primary`, `-secondary`, `-warning`, etc).

The pseudo-random color label fits better the case of random label texts (as with the Github labels, which are purely random).

> 
> > And that would require a template like this (it's a logic-less component, no ts file):
> 
> > `<span class="badge badge-{{ key }}-{{ value }}" >{{ key }}: {{ value }}</span>`
> 
> I am not really sure what are we achiving with this. Just a very basic use case for key-value pairs, we would need to have different scss classes for all the possible values? I don't see how this would really work.

Good thing of CSS is that if you don't define a class you won't get a Segfault. So we can afford creating badges with random CSS classes (`badge-service-mds`) that will never be defined. They'll simply do nothing. However, if the developer wants to highlight a specific case (`badge-service-mon` since ceph-mons are foundational to a Ceph cluster), they can do that:
```scss
.badge-service-mon {
  background-color: vv.$red-600
}
```

> After many different comments and approaches to different problems along the way I think right now we should reevaluate what we want to accomplish with this PR/component since, in my opinion we are not going anywhere with it since within every change we are trying to go with a new approach to the initial problem/solution.
> 
> Please @epuertat let me know what you think. Thank you.

I agree and I partly blame myself since I initially thought that the pseudo-random color approach was the way to go. But after reading your design doc, I realized that it wouldn't work as expected for the initial use case (service labels, 'cause we wanted to highlight only specific services), and would complicate the branding efforts. And that's been basically my proposal since then. BTW, there's nothing wrong with discarding an implementation (we do that from time to time, and it's better that to keep code that is not used).

Reply:
>No, the ngOnInit is [only triggered once](https://angular.io/guide/lifecycle-hooks) after the class constructor. If you want to respond to changes, you have to implement the ngOnChanges method. In any case, there's usually a better way to respond to changes (in fact Angular is designed to do that by default, but you have to use it correctly). Usually the best way is to expose (bind) the changing properties in the HTML template. IMHO manually tweaking the ngOnChanges is often (not always) a smell of overengineering.

Yes, I undestand that, I guess I cot confused and mixed up a couple of things, since I was testing the component at hosts page and this page uses a `cd-table` which reloads the data and creates a new table (if I am not mistaken) the `cd-label` was working fine and updating the calors approperly with the value changes.

But anyways, do you feel like we should drop the development of this component @epuertat?

Reply:
Ah, ok, if a parent component dynamically generates this, yes, that will be the case, but it's not a good practice to rely on the ngOnInit hook for things that could change after the creation of the component.

> But anyways, do you feel like we should drop the development of this component @epuertat?

I'd rather suggest to pivot/rebuild this according to the new requirements/observations, but if you prefer to close this/and open another one...